### DISCUSSION TITLE
Write logs both to stdout and to file

### DISCUSSION THREAD

**Discussion by partemyeva** [flant/shell-operator#398] [Q&A] [answered]
Hello

Is it possible to add opportunity to write logs to the file either? It seems that logrus supports logging to multiple locations from the same application: https://github.com/sirupsen/logrus/blob/b53d94c8ada4ef48988cb955795bf5c22dcf4b25/README.md#L185

The log file would be useful to configure filebeat container input for ELK integration.

Thanks

**Comment by diafour** [comment, marked as answer]
Hello! Best practice in Kubernetes is gather logs at node level, so every application just print logs to stdout.
There is no need to configure logrus, you can achieve "logging to multiple locations" using custom entrypoint.sh with tee command in it and give filebeat a try right now. 
May be you give more details? Why you need filebeat for shell-operator only?

**Reply by partemyeva** [reply]
Perhaps we don't follow the best practice, but historically each pod contains an application container and a filebeat container.

Thank you for idea with tree command!

I'm trying to use a custom entrypoint with `tee` command, my Dockerfile based on shell-operator Dockerfile:

```
FROM flant/shell-operator:v1.0.10

RUN apk add --update --no-cache \
        curl \
    rm -f /var/cache/apk/*

ENTRYPOINT ["/sbin/tini", "--", "/shell-operator"]
CMD ["start", "|", "tee", "/opt/app/log/log.json"]
```

build custom docker image `curl-shell-operator:0.1.1` and use it as container image in StatefulSet for custom operator

Here is rendered helm chart resources: [templates.zip](https://github.com/flant/shell-operator/files/8733063/templates.zip)

Then install helm chart with these resources, but an error is logged in operator container:

```
shell-operator: error: unexpected |, try --help
stream closed
```

Am I doing something wrong?

**Reply by diafour** [reply]
Yep, it's not working that way. 

Create additonal entrypoint.sh file:

```
#!/bin/bash

/shell-operator start | tee /opt/app/log/log.json
```

Add it to the image and change argument for tini:

```
FROM flant/shell-operator:v1.0.10

RUN apk add --update --no-cache \
        curl \
    rm -f /var/cache/apk/*

COPY entrypoint.sh /

ENTRYPOINT ["/sbin/tini", "--", "/entrypoint.sh"]
```

**Reply by partemyeva** [reply]
Thanks a lot for help!

Combined with `unbuffer` command (from `expect` package) it works perfect:

```
#!/bin/bash

unbuffer /shell-operator start | tee /opt/app/log/log.json
```
